In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as datetime

In [2]:
df_users = pd.read_csv('takehome_users.csv',encoding="ISO-8859-1")

In [3]:
df_engagement = pd.read_csv('takehome_user_engagement.csv',encoding="ISO-8859-1")

We'll start off by doing some traditional dataframe exploration - head, tail, describe, etc.

In [4]:
df_users.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [5]:
df_users.tail()

object_id        creation_time           name  \
11995      11996  2013-09-06 06:14:15   Meier Sophia   
11996      11997  2013-01-10 18:28:37  Fisher Amelie   
11997      11998  2014-04-27 12:45:16    Haynes Jake   
11998      11999  2012-05-31 11:55:59   Faber Annett   
11999      12000  2014-01-26 08:57:12     Lima Thaís   

                           email     creation_source  \
11995      SophiaMeier@gustr.com          ORG_INVITE   
11996     AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997        JakeHaynes@cuvox.de        GUEST_INVITE   
11998         mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999  ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN

In [6]:
df_engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [7]:
df_engagement.tail()

time_stamp  user_id  visited
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

In [8]:
df_users.describe()

object_id  last_session_creation_time  opted_in_to_mailing_list  \
count  12000.00000                8.823000e+03              12000.000000   
mean    6000.50000                1.379279e+09                  0.249500   
std     3464.24595                1.953116e+07                  0.432742   
min        1.00000                1.338452e+09                  0.000000   
25%     3000.75000                1.363195e+09                  0.000000   
50%     6000.50000                1.382888e+09                  0.000000   
75%     9000.25000                1.398443e+09                  0.000000   
max    12000.00000                1.402067e+09                  1.000000   

       enabled_for_marketing_drip        org_id  invited_by_user_id  
count                12000.000000  12000.000000         6417.000000  
mean                     0.149333    141.884583         5962.957145  
std                      0.356432    124.056723         3383.761968  
min                      0.000000      0.000000            3.000000  
25%                      0.000000     29.000000         3058.000000  
50%                      0.000000    108.000000         5954.000000  
75%                      0.000000    238.250000         8817.000000  
max                      1.000000    416.000000        11999.000000

In [9]:
df_engagement.describe()

user_id   visited
count  207917.000000  207917.0
mean     5913.314197       1.0
std      3394.941674       0.0
min         1.000000       1.0
25%      3087.000000       1.0
50%      5682.000000       1.0
75%      8944.000000       1.0
max     12000.000000       1.0

In [10]:
df_users.isnull().sum()

object_id                        0
creation_time                    0
name                             0
email                            0
creation_source                  0
last_session_creation_time    3177
opted_in_to_mailing_list         0
enabled_for_marketing_drip       0
org_id                           0
invited_by_user_id            5583
dtype: int64

We can see that there are 3,177 misisng values in the last session column and 5,583 missing values in the invited by column.

From our data dictionary we can intuit these - 3,177 individuals have not logged in, and 5583 indiviudals created an account without a specific invite from another individual. We can replace both of these with 0s.

In [11]:
df_users['last_session_creation_time'].fillna(0,inplace=True)
df_users['invited_by_user_id'].fillna(0,inplace=True)

In [13]:
df_engagement.dtypes

time_stamp    object
user_id        int64
visited        int64
dtype: object

For the purposes of feature engineering its easier if the time columns were in datetime format. We'll switch those now.

In [15]:
df_engagement['time_stamp'] = pd.to_datetime(df_engagement['time_stamp'])

We'll then create a multiindexed version of the engagement dataframe to make it easier to manipulate and do the calculations for whether a user is "adopted."

In [16]:
df_engagement_multiindex = df_engagement.set_index(['user_id','time_stamp'])

In [17]:
df_users_test = df_users

Essentially, the below calculation is looking at 7-day periods following each login to see if the subsequent two logins fall in that period (i.e., whether the user is adopted). If there are fewer than three logins, the user cannot be adopted. Because this only positively identifies adopted users but does not negatively identify users who have logged in more than three times but none within a 7 day period, we will need to fill in nans afterwards. 

In [18]:
for user_id, new_df in df_engagement_multiindex.groupby(level=0):
    if len(new_df) < 3:
        df_users_test.loc[df_users_test['object_id'] == user_id, 'adopted'] = 0
        continue
    else:
        for obj in range(len(new_df)-1):
            if(pd.to_datetime(new_df.index.values[obj][0]) + datetime.timedelta(days=7)) > pd.to_datetime(new_df.index.values[obj+2][0]):
                df_users_test.loc[df_users_test['object_id'] == user_id, 'adopted'] = 1
                break

In [19]:
df_users_test['adopted'].fillna(0,inplace=True)

The below is the start of some feature engineering for us - taking login information and translating that into logins per week and total logins.

In [20]:
for user_id, new_df in df_engagement_multiindex.groupby(level=0):
    if len(new_df) >= 2:
        df_users_test.loc[df_users_test['object_id'] == user_id, 'logins_per_week'] = len(new_df) / ((pd.to_datetime(new_df.index.values[-1][1]) - pd.to_datetime(new_df.index.values[0][1])).days / 7)
        df_users_test.loc[df_users_test['object_id'] == user_id, 'total_logins'] = len(new_df)
        continue
    else:
        df_users_test.loc[df_users_test['object_id'] == user_id, 'total_logins'] = len(new_df)  
        continue

In [21]:
df_users_test['logins_per_week'].fillna(0,inplace=True)

In [22]:
df_users_test['logins_per_week'].isnull().sum()

0

In [23]:
df_users_test['total_logins'].fillna(0,inplace=True)

In [24]:
df_users_test['adopted'].isnull().sum()

0

Now we'll drop some of the features that will not contribute to our modeling - name, email, user id, etc. We drop the creation time logs as we will not be going that much in depth in those for feature engineering. Future analyses could look at e.g. whether accounts were created on teh weekday or weekend, morning or nighttime logins, etc., but we will not do that here.

In [26]:
df_users_test = df_users_test.drop(['object_id','name','email','creation_time','last_session_creation_time'],axis=1)

KeyError: "['object_id' 'name' 'email' 'creation_time' 'last_session_creation_time'] not found in axis"

In [27]:
df_users_test.dtypes

creation_source                object
opted_in_to_mailing_list        int64
enabled_for_marketing_drip      int64
org_id                          int64
invited_by_user_id            float64
adopted                       float64
total_logins                  float64
logins_per_week               float64
dtype: object

In [28]:
df_users_test.isnull().sum()

creation_source               0
opted_in_to_mailing_list      0
enabled_for_marketing_drip    0
org_id                        0
invited_by_user_id            0
adopted                       0
total_logins                  0
logins_per_week               0
dtype: int64

In [29]:
df_users_test['invited_by_user_id'].fillna(0,inplace=True)

In [30]:
df_users_test['adopted'].value_counts()

0.0    9752
1.0    2248
Name: adopted, dtype: int64

We had this number previously, but we can see that 2248 of the 9752 users in the dataset are labeled as "adopted" based on the criteria.

In [31]:
df_users_test['total_logins'].sum()

207917.0

In [32]:
df_users_test['logins_per_week'].sum()

5000.433711916607

We'll now want to encode the total logins/logins per week variables. It makes sense that these would correlate strongly with adoption - folks who log in more often will be more likely to log in 3 times within a 7 day period - so we'll normalize them to identify relative login activity.

In [36]:
from sklearn.preprocessing import StandardScaler

In [37]:
scaler = StandardScaler()

In [38]:
df_users_test[['total_logins','logins_per_week']] = scaler.fit_transform(df_users_test[['total_logins','logins_per_week']])

While our org id is numerical, it is a categorical variable. We will want to encode it appropriately to give it the right weight in our model. Same for the invited by user ID field.

In [39]:
df_users_test['org_id'] = df_users_test['org_id'].map(str)

In [40]:
df_users_test['invited_by_user_id'] = df_users_test['invited_by_user_id'].fillna(0).map(str)

In [41]:
df_users_test

creation_source  opted_in_to_mailing_list  \
0            GUEST_INVITE                         1   
1              ORG_INVITE                         0   
2              ORG_INVITE                         0   
3            GUEST_INVITE                         0   
4            GUEST_INVITE                         0   
...                   ...                       ...   
11995          ORG_INVITE                         0   
11996  SIGNUP_GOOGLE_AUTH                         0   
11997        GUEST_INVITE                         1   
11998   PERSONAL_PROJECTS                         0   
11999              SIGNUP                         0   

       enabled_for_marketing_drip org_id invited_by_user_id  adopted  \
0                               0     11            10803.0      0.0   
1                               0      1              316.0      1.0   
2                               0     94             1525.0      0.0   
3                               0      1             5151.0      0.0   
4                               0    193             5240.0      0.0   
...                           ...    ...                ...      ...   
11995                           0     89             8263.0      0.0   
11996                           0    200                0.0      0.0   
11997                           1     83             8074.0      0.0   
11998                           0      6                0.0      0.0   
11999                           1      0                0.0      0.0   

       total_logins  logins_per_week  
0         -0.253969        -0.382683  
1         -0.051745         0.279076  
2         -0.253969        -0.382683  
3         -0.253969        -0.382683  
4         -0.253969        -0.382683  
...             ...              ...  
11995     -0.253969        -0.382683  
11996     -0.253969        -0.382683  
11997     -0.253969        -0.382683  
11998     -0.253969        -0.382683  
11999     -0.253969        -0.382683  

[12000 rows x 8 columns]

In [42]:
df_users_test.describe()

opted_in_to_mailing_list  enabled_for_marketing_drip       adopted  \
count              12000.000000                12000.000000  12000.000000   
mean                   0.249500                    0.149333      0.187333   
std                    0.432742                    0.356432      0.390195   
min                    0.000000                    0.000000      0.000000   
25%                    0.000000                    0.000000      0.000000   
50%                    0.000000                    0.000000      0.000000   
75%                    0.000000                    0.000000      0.000000   
max                    1.000000                    1.000000      1.000000   

       total_logins  logins_per_week  
count  1.200000e+04     1.200000e+04  
mean  -5.807577e-16     7.124209e-16  
std    1.000042e+00     1.000042e+00  
min   -2.695244e-01    -3.826829e-01  
25%   -2.695244e-01    -3.826829e-01  
50%   -2.539687e-01    -3.826829e-01  
75%   -2.539687e-01    -3.826829e-01  
max    9.157224e+00     1.247435e+01

Now we'll use pandas' get_dummies feature to encode our categorical features.

In [43]:
df_users_dummies = pd.get_dummies(df_users_test)

Now we'll split into predictor and response variables. Because we are not trying to build a generalizable model, just looking at feature importance, we don't need to cross validate our model on a training set.

In [49]:
X = df_users_dummies.drop('adopted',axis=1)

In [50]:
y = df_users_dummies['adopted']

In [51]:
from sklearn.ensemble import RandomForestClassifier

In [52]:
forest = RandomForestClassifier()

In [53]:
forest.fit(X,y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [54]:
feature_imp = list(zip(X.columns,forest.feature_importances_))

In [55]:
feature_imp.sort(key=lambda x: x[1],reverse=True)

In [56]:
feature_imp[:10]

[('total_logins', 0.5020274279739768),
 ('logins_per_week', 0.4118714827513109),
 ('creation_source_PERSONAL_PROJECTS', 0.0028312071684989263),
 ('opted_in_to_mailing_list', 0.002238595594604685),
 ('creation_source_GUEST_INVITE', 0.0015951972642529882),
 ('enabled_for_marketing_drip', 0.0014310471070300842),
 ('creation_source_SIGNUP_GOOGLE_AUTH', 0.0010274355526841328),
 ('creation_source_ORG_INVITE', 0.0009880007237505087),
 ('org_id_0', 0.0008799304563266245),
 ('creation_source_SIGNUP', 0.0007954579637396096)]

Here we can see the ten most "predictive" variables. We see at the top of the charts the engineered features, which doesn't tell us much as they are both related to the magnitude of login activity for the user. We can see that the personal projects invite is the next highest predictive indicator, 